python-threatexchange/threatexchange/cli/dataset_cmd.py

```diff
@@ -1,184 +1,256 @@
         print_records: bool,
+        # Signal selectors
+        only_collabs: t.List[str],
+        only_signals: t.List[t.Type[SignalType]],
+        only_content: t.List[t.Type[ContentType]],
+        only_tags: t.List[str],
+        # Print stuff
+        signals_only: bool,
+        limit: t.Optional[int],
     ) -> None:
+        self.clear_indices = clear_indices
         self.rebuild_indices = rebuild_indices
-        self.only_type = only_type
-        self.only_tag = only_tag
-        self.indicator_only = indicator_only
-        self.signal_summary = signal_summary
         self.print_records = print_records
+        self.signal_summary = signal_summary
 
-    def execute(self, api: ThreatExchangeAPI, dataset: Dataset) -> None:
-        stores = [
-            threat_updates.ThreatUpdateFileStore(
-                dataset.state_dir,
-                privacy_group,
-                api.app_id,
-                serialization=CliIndicatorSerialization,
-            )
-            for privacy_group in dataset.config.privacy_groups
-        ]
-        indicators = {}
-        for store in stores:
-            indicators.update(store.load_state())
-
-        if self.only_type:
-            signal_types = []
-            s_type = meta.get_signal_types_by_name().get(self.only_type)
-            if s_type:
-                signal_types.append(s_type)
-                self.signal_summary = True
-
-            content_type = meta.get_content_types_by_name().get(self.only_type)
-            if content_type:
-                signal_types.extend(content_type.get_signal_types())
-                self.signal_summary = True
-
-            indicators = {
-                k: v
-                for k, v in indicators.items()
-                if v.indicator_type == self.only_type
-                or (
-                    signal_types
-                    and any(
-                        sig_type.indicator_applies(v.indicator_type, v.rollup.labels)
-                        for sig_type in signal_types
-                    )
-                )
-            }
-
-        if self.only_tag:
-            self.signal_summary = True
-            indicators = {
-                k: v for k, v in indicators.items() if self.only_tag in v.rollup.labels
-            }
-
-        if self.rebuild_indices:
-            generate_cli_indices(dataset, stores)
-        if self.print_records:
-            self._print_records(indicators)
+        self.only_collabs = set(only_collabs)
+        self.only_signals = set(only_signals)
+        self.only_content = set(only_content)
+        self.only_tags = set(only_tags)
+        self.signals_only = signals_only
+        self.limit = limit
+
+    def execute(self, settings: CLISettings) -> None:
+        # Maybe consider subcommands?
+        if self.clear_indices:
+            self.execute_clear_indices(settings)
+        elif self.rebuild_indices:
+            self.execute_generate_indices(settings)
```

Comment:
clear before generating?

Reply:
I'll clear them inline so that you don't end up with no indices if it errors